source/tests/unit/session_resource_repository_tests.cpp

```diff
@@ -0,0 +1,259 @@
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <server/session_repository.h>
+#include <server/session_resource_repository.h>
+
+using namespace nidevice_grpc;
+
+namespace ni {
+namespace tests {
+namespace unit {
+
+const uint32_t kNoSession = 0;
+const int32_t kNoError = 0;
+
+template <typename TResourceHandle>
+uint32_t add_session_resource(
+    SessionResourceRepository<TResourceHandle>& resource_repository,
+    const TResourceHandle& resource)
+{
+  uint32_t session_id;
+  auto result = resource_repository.add_session(
+      "",
+      [=]() { return std::make_tuple(kNoError, resource); },
+      [](TResourceHandle handle) {},
+      session_id);
+  EXPECT_EQ(kNoError, result);
+  return session_id;
+}
+
+TEST(SessionResourceRepositoryTests, AddSessionResource_ResourceIsAdded)
+{
+  SessionRepository repository;
+  ;
+  SessionResourceRepository<uint64_t> resource_repository(
+      &repository);
+
+  const uint64_t kResourceHandle = 0x1111222233334444;
+  const std::string kSessionName("session");
+  uint32_t session_id;
+  auto result = resource_repository.add_session(
+      kSessionName,
+      [=]() { return std::make_tuple(kNoError, kResourceHandle); },
+      [](uint64_t handle) {},
+      session_id);
+
+  EXPECT_EQ(kNoError, result);
+  EXPECT_NE(kNoSession, session_id);
+  EXPECT_EQ(
+      kResourceHandle,
+      resource_repository.access_session(session_id, ""));
+  EXPECT_EQ(
+      kResourceHandle,
+      resource_repository.access_session(0, kSessionName));
+  EXPECT_EQ(
+      session_id,
+      repository.access_session(0, kSessionName));
+}
+
+TEST(SessionResourceRepositoryTests, SessionResource_RemoveSessionResource_ResourceIsRemoved)
+{
+  SessionRepository repository;
+  ;
+  SessionResourceRepository<uint16_t> resource_repository(
+      &repository);
+  const uint16_t kResourceHandle = 0x1234;
+  auto session_id = add_session_resource(resource_repository, kResourceHandle);
+
+  resource_repository.remove_session(kResourceHandle);
+
+  EXPECT_EQ(
+      kNoSession,
+      resource_repository.access_session(session_id, ""));
+  EXPECT_EQ(
+      kNoSession,
+      repository.access_session(session_id, ""));
+}
+
+TEST(SessionResourceRepositoryTests, SessionResource_RemoveFromResourceRepository_ResourceIsRemoved)
```

Comment:
I might be missing something but is this test any different than `SessionResource_RemoveSessionResource_ResourceIsRemoved` besides the extra assert on `resolve_session_id`.

Reply:
I mistyped the test. It should be deleting the session_id from the session_repository. So we test remove both ways.